### PR TITLE
feat(identity-rs): store and fetch did documents from dht

### PR DESCRIPTION
**Pull request by MatejVukosav**
- store and retrieve did document from provided dht kademlia

Output
```
Generating identity
Stored did document: {
  "id": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ",
  "verificationMethod": [
    {
      "id": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ#key1",
      "type": "Ed25519",
      "publicKeyMultibase": "zCovLVG4fQcqSmSqox5oVUdK4pvMsuNxbZLLs6gwwBrqqjFdz3QFGTsgtopjX91ek2gk2SkQ",
      "controller": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ"
    }
  ]
}
Fetched did document {
  "id": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ",
  "verificationMethod": [
    {
      "id": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ#key1",
      "type": "Ed25519",
      "publicKeyMultibase": "zCovLVG4fQcqSmSqox5oVUdK4pvMsuNxbZLLs6gwwBrqqjFdz3QFGTsgtopjX91ek2gk2SkQ",
      "controller": "did:cali:12D3KooWK2kSHzeTE5daWqFtqN2GMoUNFFN6Lps8zz47hHy5hAhQ"
    }
  ]
}
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [workspace]
 resolver = "1"
-members = ["./chat-p0c"]
+members = ["./chat-p0c", "./identity-rs"]
```
Comment (miraclx) [inline Cargo.toml]: 👍🏽

File: identity-rs/.gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 .idea/
 data/
 target/
+.DS_Store
```
Comment (miraclx) [inline identity-rs/.gitignore]: Add to the `.gitignore` at the root instead

File: identity-rs/src/dht.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+use libp2p::kad::store::MemoryStore;
+use libp2p::kad::store::RecordStore;
+
+use crate::types::DidDocument;
+
+pub struct Dht<'a> {
+    kad: &'a mut libp2p::kad::store::MemoryStore,
```
Comment (miraclx) [inline identity-rs/src/dht.rs]: If we intend to merge this with the networking stack, then ideally, these write operations will need to be broadcasted to the connected peers, in which case, you should instead be using [`kad::Behaviour::put_record`](https://docs.rs/libp2p/0.53.2/libp2p/kad/struct.Behaviour.html#method.put_record) to do this.

But before this integration, I suppose this is fine.

File: identity-rs/Cargo.toml [hunk 4]
```diff
@@ -7,13 +7,13 @@ edition = "2021"
 
 [dependencies]
 didkit = "0.6.0"
-libp2p = "0.53.2"
+libp2p = { workspace = true, features = ["full"] }
```
Comment (miraclx) [inline identity-rs/Cargo.toml]: Looks like with the reduced scope, you may only need the `kad` feature.

```suggestion
libp2p = { workspace = true, features = ["kad"] }
```

I'll be reducing the feature set of the networking stack as well once we've gotten all we need.

File: identity-rs/src/dht.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        let key = libp2p::kad::RecordKey::from(key_id);
+        if let Some(result) = self.kad.get(&key) {
+            let value = &result.value.clone();
+            let data = String::from_utf8(value.to_vec()).ok()?;
+            let did_document: DidDocument =
+                serde_json::from_str(&data).expect("Failed to parse JSON into DidDocument");
+            return Some(did_document);
+        } else {
+            return None;
+        }
```
Comment (miraclx) [inline identity-rs/src/dht.rs]: `?` to shortcircuit

```suggestion
        let result = self.kad.get(&key)?;
        let value = &result.value.clone();
        let data = String::from_utf8(value.to_vec()).ok()?;
        let did_document: DidDocument =
            serde_json::from_str(&data).expect("Failed to parse JSON into DidDocument");
        Some(did_document)
```

File: identity-rs/src/dht.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    pub fn new(kad: &'a mut MemoryStore) -> Self {
+        Dht { kad }
+    }
+
+    /// Write did in dht
+    pub fn write_record(&mut self, did_document: DidDocument) {
+        let key_id: Vec<u8> = did_document.clone().id.into();
+        let key = libp2p::kad::RecordKey::from(key_id);
+        let value: Vec<u8> =
+            serde_json::to_vec(&did_document.clone()).expect("Serialization failed");
```
Comment (miraclx) [inline identity-rs/src/dht.rs]: Return a result instead of panicking.

File: identity-rs/src/dht.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    /// read did document per did id from dht
+    pub fn read_record(&self, did: String) -> Option<DidDocument> {
+        let key_id: Vec<u8> = did.into();
+        let key = libp2p::kad::RecordKey::from(key_id);
+        if let Some(result) = self.kad.get(&key) {
+            let value = &result.value.clone();
+            let data = String::from_utf8(value.to_vec()).ok()?;
+            let did_document: DidDocument =
+                serde_json::from_str(&data).expect("Failed to parse JSON into DidDocument");
```
Comment (miraclx) [inline identity-rs/src/dht.rs]: Same here

File: identity-rs/src/identity_provider.rs [hunk 8]
```diff
@@ -41,14 +44,18 @@ pub fn create_identity(authentication: Authentication) -> DidDocument {
...
-#[allow(dead_code)]
-pub async fn get_identifier(_did: String) -> Option<DidDocument> {
-    //TODO fetch document per key from storage
-    unimplemented!();
+pub fn get_identifier(store: &mut MemoryStore, did: String) -> Option<DidDocument> {
+    if let Some(did_document) = Dht::new(store).read_record(did.clone()) {
+        return Some(did_document);
+    } else {
+        return None;
+    }
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]:
```suggestion
    Dht::new(store).read_record(did)
```

File: identity-rs/src/types.rs [hunk 9]
```diff
@@ -1,10 +1,10 @@
-use std::fmt;
 use serde::{Deserialize, Serialize};
+use std::fmt;
```
Comment (miraclx) [inline identity-rs/src/types.rs]: Let's order imports akin to the convention in https://github.com/calimero-is-near/cali2.0-experimental/blob/14e643e4ebd5d6b755f5f5290c3f77620c706c43/chat-p0c/src/network/mod.rs#L1-L24

1. `std` imports
2. third-party crates imports
3. self-crate imports
4. child module imports

```suggestion
use std::fmt;

use serde::{Deserialize, Serialize};
```

File: identity-rs/src/main.rs [hunk 10]
```diff
@@ -1,11 +1,14 @@
...
 
+mod dht;
 mod identity_provider;
 mod types;
 
-use libp2p::identity::Keypair;
+use libp2p::{identity::Keypair, kad::store::MemoryStore};
 
-use crate::types::AlgorithmType;
+use crate::{identity_provider::get_identifier, types::AlgorithmType};
```
Comment (miraclx) [inline identity-rs/src/main.rs]:
```suggestion
use std::borrow::BorrowMut;

use libp2p::{identity::Keypair, kad::store::MemoryStore};

mod dht;
mod identity_provider;
mod types;

use identity_provider::{create_identity, get_identifier, Authentication};
use types::AlgorithmType;
```

File: identity-rs/Cargo.toml (hunk 4 above)
Comment (MatejVukosav) [reply identity-rs/Cargo.toml]: I'll leave full scope for now and reduce it once we have everything functional.

File: identity-rs/src/dht.rs (hunk 6 above)
Comment (MatejVukosav) [reply identity-rs/src/dht.rs]: Add ->Result<(), io::Error>

File: identity-rs/src/dht.rs (hunk 5 above)
Comment (MatejVukosav) [reply identity-rs/src/dht.rs]: ? can only be used on Result and `self.kad.get(&key)` returns `Option`

File: identity-rs/src/types.rs (hunk 9 above)
Comment (MatejVukosav) [reply identity-rs/src/types.rs]: Do you have any custom formatter for that case cause standard fmt does not change it by default?

File: identity-rs/src/dht.rs (hunk 3 above)
Comment (MatejVukosav) [reply identity-rs/src/dht.rs]: I would prefer to have an interface in identity for Kad and not to care how it is stored. Ideally from shared lib but that is something out of my scope right now.

File: identity-rs/src/dht.rs [hunk 11]
```diff
@@ -0,0 +1,44 @@
+use std::io;
+
+use libp2p::kad::store::MemoryStore;
+use libp2p::kad::store::RecordStore;
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
use libp2p::kad::store::{MemoryStore, RecordStore};
```

File: identity-rs/src/dht.rs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+use crate::types::DidDocument;
+
+pub struct Dht<'a> {
+    kad: &'a mut libp2p::kad::store::MemoryStore,
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
    kad: &'a mut MemoryStore,
```

File: identity-rs/src/dht.rs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+impl<'a> Dht<'a> {
+    pub fn new(kad: &'a mut MemoryStore) -> Self {
+        Dht { kad }
+    }
+
+    /// Write did in dht
+    pub fn write_record(&mut self, did_document: DidDocument) -> Result<(), io::Error> {
+        let key_id: Vec<u8> = did_document.clone().id.into();
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
        let key_id = did_document.id.as_bytes().to_owned();
```

File: identity-rs/src/dht.rs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        let key = libp2p::kad::RecordKey::from(key_id);
+        let value: Vec<u8> = serde_json::to_vec(&did_document.clone())?;
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
        let value = serde_json::to_vec(&did_document)?;
```

File: identity-rs/src/dht.rs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        let record = libp2p::kad::Record::new(key, value.clone());
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
        let record = libp2p::kad::Record::new(key, value);
```

File: identity-rs/src/dht.rs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    /// read did document per did id from dht
+    pub fn read_record(&self, did: String) -> Result<Option<DidDocument>, io::Error> {
+        let key_id: Vec<u8> = did.into();
+        let key = libp2p::kad::RecordKey::from(key_id);
+
+        if let Some(result) = self.kad.get(&key) {
+            let value = &result.value.clone();
+            let data = String::from_utf8(value.to_vec())
+                .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+            let did_document: DidDocument = serde_json::from_str(&data)?;
```
Comment (miraclx) [inline identity-rs/src/dht.rs]:
```suggestion
            let data = std::str::from_utf8(&result.value)
                .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
            let did_document: DidDocument = serde_json::from_str(data)?;
```

File: identity-rs/src/types.rs (hunk 9 above)
Comment (miraclx) [reply identity-rs/src/types.rs]: `rustfmt` won't merge them if you have spaces in between those sections

Here's a pattern:

```rust
// libstd
use std::fmt;
use std::io;

// crate imports
use color_eyre::eyre;
use serde::{Serialize, Deserialize};
use tracing::{info, log};

// crate imports
use crate::root_module;

// child module declarations
mod types;
mod utils;

// child module imports
use utils::do_something;
```

**Pull request merged**